parser/src/ast.rs

```diff
@@ -0,0 +1,101 @@
+use crate::lexer::Token;
+use std::fmt;
+
+/// [Identifier] is used to represent variable names.
+#[derive(Debug, Eq, PartialEq)]
+pub struct Identifier {
+    pub name: String,
+}
+
+impl fmt::Display for Identifier {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", &self.name)
+    }
+}
+
+/// [Source] is the root node of the AST representing the AIR constraints file.
+#[derive(Debug, PartialEq)]
+pub struct Source(pub Vec<SourceSection>);
+
+/// Source is divided into SourceSections.
+/// There are 4 types of Source Sections:
+/// - AirDef: Name of the air constraints module.
+/// - TraceCols: Trace Columns representing columns of the execution trace.
+/// - BoundaryConstraints: Boundary Constraints to be enforced on the boundaries of columns defined
+///   in the TraceCols section. Currently there are two types of boundaries, First and Last
+///   representing the first and last rows of the column.
+/// - TransitionConstraints: Transition Constraints to be enforced on the trace columns defined
+///   in the TraceCols section.
+#[derive(Debug, PartialEq)]
+pub enum SourceSection {
+    AirDef(AirDef),
+    TraceCols(TraceCols),
+    BoundaryConstraints(BoundaryConstraints),
+    TransitionConstraints(TransitionConstraints),
+}
+
+/// Name of the air constraints module.
+#[derive(Debug, Eq, PartialEq)]
+pub struct AirDef {
+    pub name: Identifier,
+}
+
+/// [TraceCols] contains the trace columns representing columns of the execution trace.
+/// TraceCols is made up of TraceColsGrp which represents groups of columns of a specific type.
+#[derive(Debug, PartialEq)]
+pub struct TraceCols {
+    pub cols: Vec<TraceColsGrp>,
+}
+
+/// [TraceColsGrp] represents a group of columns of a specfic type.
+/// There are two types of groups:
+/// - MainTraceCols
+/// - AuxiliaryTraceCols
+#[derive(Debug, PartialEq)]
+pub struct TraceColsGrp {
+    pub cols_grp_type: TraceColsGrpType,
+    pub cols: Vec<Expr>,
+}
+
+/// Describes the type of trace columns group.
+#[derive(Debug, Eq, PartialEq)]
+pub enum TraceColsGrpType {
+    MainTraceCols,
+    AuxiliaryTraceCols,
+}
+
+/// Stores the boundary constraints to be enforced to the trace column values.
+#[derive(Debug, PartialEq)]
+pub struct BoundaryConstraints {
+    pub boundary_constraints: Vec<Constraint>,
+}
+
+/// Describes the type of boundary in the boundary constraint.
+#[derive(Debug, Eq, PartialEq)]
+pub enum Boundary {
+    First,
+    Last,
+}
+
+/// Stores the transition constraints to be enforced to the trace column values.
+#[derive(Debug, PartialEq)]
+pub struct TransitionConstraints {
+    pub transition_constraints: Vec<Constraint>,
+}
+
+/// Stores the expression corresponding to the transition or boundary constraint.
+#[derive(Debug, PartialEq)]
+pub struct Constraint {
+    pub expr: Expr,
```

Comment:
I think this might be too permissive. A constraint must always contain an equality expression. For example: `a = b`. Should this be defined as:
```
pub struct Constraint {
    pub lhs: Expr,
    pub rhs: Expr,
}
```

Also, should there be different types for transition vs. boundary constraints? For example, `enf a.first = b` is a valid boundary constraint, but is not a valid transition constraint. Or will we do these types of checks later on?

Reply:
I've changed the constraint definition to what you suggested to make it more restrictive.

> Also, should there be different types for transition vs. boundary constraints? For example, enf a.first = b is a valid boundary constraint, but is not a valid transition constraint. Or will we do these types of checks later on?

I think it might be better to handle it after the AST is built otherwise it seems like the grammar would become much more complex (unless I'm missing something?) and it will be easier to do these checks once the AST is built. I'll think on this more though.

Reply:
> I think it might be better to handle it after the AST is built otherwise it seems like the grammar would become much more complex

Agreed! Let's do these checks later.

Reply:
Actually, thinking about it some more, I wonder if it is actually not too complicated. Here is how it might work:

We could change the current `Constraint` to `TransitionConstraint` and it would look like so:
```Rust
pub struct TransitionConstraint {
    pub lhs: Expr,
    pub rhs: Expr,
}
```

We could add a new struct for boundary constraint which could look something like:
```Rust
pub struct BoundaryConstraint {
    pub column: Identifier,
    pub boundary: Boundary,
    pub value: Expr,
}
```

Then, `TransitionConstraints` and `BoundaryConstraints` would look like so:
```Rust
pub struct TransitionConstraints {
    pub transition_constraints: Vec<TransitionConstraint>,
}

pub struct BoundaryConstraints {
    pub boundary_constraints: Vec<BoundaryConstraints>,
}
```
And finally, we would be able to remove `Boundary` variant from the `Expr` enum.

Do you think this would complicate things too much?

Reply:
Yeah, sorry, I was mistaken, this will work. I have changed it to what you suggested.